lib/opencensus-proto/src/span_exporter.rs

```
@@ -0,0 +1,140 @@
+    /// the streaming request or if there isn't a span available in the spans
+    /// stream.  Returns Ready if a span was transfered successfully.  Returns
+    /// Err if any operation fails, which indicates a new streaming request
+    /// must be established.
+    fn poll_send_span(
+        &mut self,
+        sender: &mut mpsc::Sender<ExportTraceServiceRequest>,
+        sent_node: bool,
+    ) -> Poll<(), ()> {
+        try_ready!(sender.poll_ready().map_err(|_| ()));
+        let span = try_ready!(self.spans.poll().map_err(|_| ()))
+            .expect("span stream should never terminate");
+        let node = if sent_node {
+            None
+        } else {
+            Some(self.node.clone())
+        };
+        let msg = ExportTraceServiceRequest {
+            node,
+            spans: vec![span],
+            resource: None,
+        };
+        trace!("transmitting span: {:?}", msg);
+        sender
+            .try_send(msg)
+            .map(|()| Async::Ready(()))
+            .map_err(|e| debug!("failed to transmit span: {}", e))
+    }
+}
+
+impl<T, S> Future for SpanExporter<T, S>
+where
+    T: GrpcService<BoxBody>,
+    S: Stream<Item = Span>,
+{
+    type Item = ();
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        loop {
+            let state = match self.state.take().expect("corrupt state") {
+                State::Idle => {
+                    // If the request stream fails, all spans in this buffer
+                    // will be lost.  Therefore, we keep this buffer small to
+                    // minimize span loss.  Keeping this buffer small is okay
+                    // because we will exert backpressure and can buffer spans
+                    // elsewhere upstream.
+                    // TODO: Does this make sense?
+                    let (tx, rx) = mpsc::channel(1);
+                    let mut svc = TraceService::new(self.client.as_service());
```

I have some suggestions to simplify this a bit, but there's a glaring hole here: We need to call `poll_ready` on the client before sending a message here. What do you propose should be the behavior if poll_ready fails (indicating the client is now unusable)?

---

under what conditions will the client's poll_ready return an error?  if the client has a reconnect module, can we assume this won't happen and simply fail this future if it does?

---

Yeah, on further thought, I think it's okay for the span_exporter to handle errors fatally, but we should probably propagate the error and not swallow it in the future.